### PR TITLE
Fixed an issue where there was a delay when deleting items

### LINKED ISSUE #2672: Lag in more than 10 seconds between pressing delete button and appearing dialogue window
**Describe the bug**
When I'm trying to delete any files in Files app dialogue window appears to confirm the deletion (which in itself is not good, Explorer allows you to turn off confirmation for moving to recycle bin) but the biggest problem is that it appears with a terrible lag of more than 10 seconds.

**To Reproduce**
1. Go to any directory
2. Select any file or folder
3. Press Del button on the keyboard or choose  Delete from context menu
4. Wait

**Expected behavior**
There should be no delay 

**Desktop (please complete the following information):**
 - OS Version: [e.g. Windows 10 2004 20279 fe]
 - App version: [v0.23.1]

**Log file**
Please post the log file here so that we can understand your problem better. You can access it from Settings->About->Open log location.

### LINKED ISSUE #2732: Small files taking 6 whole seconds to delete 😒
**Describe the bug**
With a recent update, when you press the delete button, it takes about 6 whole seconds before the file gets deleted. This is quite irritating, please look into this.

**To Reproduce**
1. Disable the deletion confirmation (Settings > Preferences)
2. Locate any file.
3. Press the delete button.

**Expected behavior**
I expected the small text file to disappear almost instantly when deleted, like it used to before the update.

**Screenshots**
N/A

**Desktop (please complete the following information):**
 - OS Version: Windows 10 Ver. 20H2 (OS Build 19042.685)
 - App version: [v0.23.1.0]

**Additional context**
After restarting and re-installing the app, the issue is still present.

**Log file**
Debug:
2020-12-27 15:28:57.5544|INFO|.ctor|App launched|
2020-12-27 15:28:58.0792|INFO|.ctor|Drive added: C:\, Fixed|
2020-12-27 15:28:58.1832|INFO|.ctor|Drive added: D:\, Fixed|
2020-12-27 15:28:58.2831|INFO|.ctor|Drive added: E:\, CDRom|

Debug Full Trust:
2020-12-27 15:28:58.7420|INFO|FilesFullTrust.QuickLook|QuickLook detected: False|
2020-12-27 15:29:29.4373|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.5382|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.5532|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.5712|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.5802|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.5942|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6092|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6412|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6412|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6602|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6722|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6722|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6872|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.6872|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.7062|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:29.7202|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.0631|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.1021|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.1261|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.1531|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.1711|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.2191|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:30.5010|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.8121|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.8380|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.8595|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.8815|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.9005|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.9255|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:31.9810|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:32.0050|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:32.0290|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:32.0660|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:32.0880|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:32.1090|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:33.7170|INFO|FilesFullTrust.Program|Argument: RecycleBin|
2020-12-27 15:29:39.7186|INFO|FilesFullTrust.Program|Argument: RecycleBin|
2020-12-27 15:29:40.5893|INFO|FilesFullTrust.Program|Argument: RecycleBin|
2020-12-27 15:29:41.5005|INFO|FilesFullTrust.Program|Argument: RecycleBin|
2020-12-27 15:29:47.5791|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.5135|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.5325|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.5525|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.5725|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.5915|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.6105|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.6294|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.6474|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.6634|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.6834|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.7024|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|
2020-12-27 15:29:49.7204|INFO|FilesFullTrust.Program|Argument: GetIconOverlay|

### PR DESCRIPTION
**Pull request by Don-Vito**
Fixes #2672
Fixes #2732

There is a significant lag on deletion, especially when numerous files are deleted.
This is significantly impacted by checking whether the file is in Recycle Bin.

This PR somewhat helps by:
* Delaying the check until the dialog presentation
(resolves the issue for those who disabled the dialog)
* Instead of searching in recycle bin just use path to decide
if the item is under recycle bin

### CODE REVIEW

**Comment by gave92** [comment]
Thanks a lot for this. I'm not sure *why* we check if we are deleting a file from recycle bin that way when we could just check of the path starts with "X:\$Recycle.Bin"..

**Comment by Don-Vito** [comment]
@gave92 - agree. Do you want me to implement this?

**Comment by gave92** [comment]
Yes please! (Obvious note: the file path of a recycle bin element can start with any drive letter, not just "C:\\"=

**Comment by gave92** [comment]
The same happens in FilesystemOperations.cs

**Comment by Don-Vito** [comment]
> `IsRecycleBinItem`

I prefer to introduce a new helper: as probably we still want to have a function that actively checks if the file is in recycle bin

**Comment by gave92** [comment]
Depending on the scope of this PR you might also optimize the call to `List<ShellFileItem> items = await recycleBinHelpers.EnumerateRecycleBin();` in `FilesystemOperations` so it's only called once when deleting many items. Not doing it is also ok.

**Comment by Don-Vito** [comment]
@gave92 - finally had a lunch time to squeeze a bit of coding :smile:
I am not sure what is the best way to approach the `List<ShellFileItem> items = await recycleBinHelpers.EnumerateRecycleBin();` as it seems that we really need shell file items there.
Probably this could be postponed.

**Comment by gave92** [comment]
@Don-Vito yup I'm didn't mean to remove it, I meant that when you delete a bunch of file that thing is called for every file, while it should be more efficient to call it once at the and

**Comment by Don-Vito** [comment]
> 
> 
> @Don-Vito yup I'm didn't mean to remove it, I meant that when you delete a bunch of file that thing is called for every file, while it would be more efficient to call it once at the end.
> It's not necessary to do this change here though

I can followup on this :blush: I will try to create an issue later on.

File: Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs [hunk 1]
```diff
@@ -179,10 +173,15 @@ public async Task<ReturnResult> DeleteItemsAsync(IEnumerable<IStorageItemWithPat
             return returnStatus;
         }
 
+        private ISet<string> getPathsUnderRecycleBin(IEnumerable<IStorageItemWithPath> source)
```
Comment (d2dyno1) [inline Files/Filesystem/FilesystemOperations/Helpers/FilesystemHelpers.cs:176]: Uppercase FunctionName()

File: Files/Helpers/RecycleBinHelpers.cs [hunk 2]
```diff
@@ -13,6 +14,7 @@ namespace Files.Helpers
     public class RecycleBinHelpers : IDisposable
     {
         #region Private Members
+        private static readonly Regex RECYCLE_BIN_PATH_REGEX = new Regex(@"\w:\\\$Recycle\.Bin\\.*");
```
Comment (d2dyno1) [inline Files/Helpers/RecycleBinHelpers.cs:17]: camelCase

**Comment by gave92** [comment]
@Don-Vito for reference: https://github.com/files-community/Files/wiki/Code-style

**Comment by Don-Vito** [comment]
@d2dyno1, @gave92 - thanks.. and sorry - didn't touch C# fort the last 8 years :)

**Comment by Don-Vito** [comment]
@gave92 - anything I can do to help pushing it forward? :angel:

**Comment by gave92** [comment]
@Don-Vito this change looks good to me, I'll approve it as soon as I can test it 👍

**Review by gave92 — APPROVED** [review]

**Pull request merged**